zkvm/docs/zkvm-impl.md

```diff
@@ -0,0 +1,76 @@
+# ZkVM implementation guide
+
+The [ZkVM specification](zkvm-spec.md) covers the _verification rules_ for transactions that are already created,
+but does not describe how to _create_ such transactions. This is the goal of this document.
+
+## Prover and Verifier APIs
+
+The [`Prover`](src/prover.rs) and [`Verifier`](src/verifier.rs) are two entry points to the ZkVM:
+
+`Prover` is an API for _creating_ a transaction object `Tx` out of a stream of `Instruction`s:
+`VM` executes the instructions, producing a serialized program bytecode, aggregated transaction signature
+and a zero-knowledge R1CS proof for the [Cloak protocol](../../spacesuit/spec.md) and custom statements expressed by instructions.
+
+`Verifier` is an API for _verifying_ a transaction object `Tx`: it parses the bytecode, executes it, verifies the aggregated transaction signature and the R1CS proof, producing a `VerifiedTx` as a result. Verification logic is described by the [ZkVM specification](zkvm-spec.md).
+
+Note that `VM` execution is used twice: first, for _proving_, then for _verifying_. 
+
+How does the `Prover` know how to sign transaction and make a proof? The prover’s input is not an opaque sequence of instruction codes, but _witness-bearing instructions_. That is, a `push` instruction on the prover’s side does not hold an opaque string of bytes, but an accurate _witness type_ that may contain secret data and necessary structure for creating the proofs and signatures.
+
+## Witness types
+
+**Witness** is a type necessary to create a zero-knowledge proof or a signature.
+
+R1CS variables’ _witness_ is their secret _assignment_.
+
+VM instructions expect different types of witness data depending on the kind of object they operate on.
+
+### Commitments
+
+For the verifier, Pedersen commitments are opaque 32-byte [compressed points](zkvm-spec.md#point). For the verifier, these are represented by a `Data::Commitment` variant that holds a `Commitment` structure which can be `::Closed` or `::Open`.
+
+The `var` instruction in the prover’s VM expects an `::Open` commitment. An open commitment contains the secret value and its blinding factor explicitly, so they can be used to compute a R1CS proof.
+
+### Predicates
+
+`Predicate` type represents either an opaque predicate, or an entire tree where unused branches can be left opaque and visited branches are stored as-is. This structure can then produce correct input for the [`call`](zkvm-spec.md#call), [`left`](zkvm-spec.md#left) and [`right`](zkvm-spec.md#right) instructions.
```

Comment:
link to spec here? for more details on the `Predicate` type

Reply:
ok!